File: kevoree-experiment/org.kevoree.experiment.smartForest/src/main/scala/org/kevoree/experiment/smartForest/dpa/AddComponentDPA.scala
package org.kevoree.experiment.smartForest.dpa

import org.kevoree.library.tools.dpa.DPA
import java.util.ArrayList
import org.kevoree.{ContainerRoot, NamedElement, ComponentType}
import scala.collection.JavaConversions._
import org.kevoree.tools.marShell.ast.TransactionalBloc._
import org.kevoree.tools.marShell.ast.AddComponentInstanceStatment._
import org.kevoree.tools.marShell.ast.ComponentInstanceID._
import org.kevoree.tools.marShell.ast.{ComponentInstanceID, AddComponentInstanceStatment, TransactionalBloc, Script}

/**
 * User: ffouquet
 * Date: 18/07/11
 * Time: 17:43
 */

object AddComponentDPAO {
  val componentName: String = "component"
  def getComponentName = componentName
  val typeDefinition: String = "type"
  def getTypeDefinition = typeDefinition
  val nodeName: String = "node"
  def getNodeName = nodeName
}

class AddComponentDPA extends DPA {


  final val templateScript: String = "tblock { \n" + "  addComponent ${component}@${node} : ${type}\n" + "}"
  private var increment: Int = 0

  def applyPointcut(myModel: ContainerRoot): java.util.List[java.util.Map[String, NamedElement]] = {
    val results: java.util.List[java.util.Map[String, NamedElement]] = new java.util.ArrayList[java.util.Map[String, NamedElement]]
    for (containerNode <- myModel.getNodes) {
      for (typeDef <- myModel.getTypeDefinitions) {
        if (typeDef.isInstanceOf[ComponentType]) {
          val myMap: java.util.Map[String, NamedElement] = new java.util.HashMap[String, NamedElement]
          myMap.put(AddComponentDPAO.typeDefinition, typeDef.asInstanceOf[NamedElement])
          myMap.put(AddComponentDPAO.nodeName, containerNode.asInstanceOf[NamedElement])
          results.add(myMap)
        }
      }
    }
    return results
  }

  def getScript(myMap: java.util.Map[String, NamedElement]): String = {
    var script: String = templateScript
    for (name <- myMap.keySet) {
      val replacedString: String = "${" + name + "}"
      script = script.replace(replacedString, myMap.get(name).getName)
    }
    script = script.replace("${" + AddComponentDPAO.componentName + "}", myMap.get(AddComponentDPAO.typeDefinition).getName + ({
      increment += 1;
      increment
    }))
    script
  }

  def getASTScript(stringNamedElementMap: java.util.Map[String, NamedElement]): Script = {
    Script(
      List(
        TransactionalBloc(
          List(
            AddComponentInstanceStatment(
              ComponentInstanceID(stringNamedElementMap.get(AddComponentDPAO.typeDefinition).getName+ ({increment += 1;increment}), Some(stringNamedElementMap.get(AddComponentDPAO.nodeName).getName)),
              stringNamedElementMap.get(AddComponentDPAO.typeDefinition).getName,
              new java.util.Properties()
            )
          )
        )
      )
    )
  }

}

File: kevoree-experiment/org.kevoree.experiment.smartForest/src/main/java/org/kevoree/library/reasoner/ecj/dpa/AddComponentDPA.java
package org.kevoree.library.reasoner.ecj.dpa;

import org.kevoree.*;
import org.kevoree.library.tools.dpa.DPA;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AddComponentDPA implements DPA {

    public final static String componentName = "component";
    public final static String typeDefinition = "type";
    public final static String nodeName = "node";

    public final static String templateScript = "tblock { \n" +
            "  addComponent ${component}@${node} : ${type}\n" +
            "}";
    private static int increment = 0;

    public List<Map<String, NamedElement>> applyPointcut(ContainerRoot myModel) {
        List<Map<String, NamedElement>> results = new ArrayList();
        for (ContainerNode containerNode : myModel.getNodes()) {
            for (TypeDefinition typeDef : myModel.getTypeDefinitions()) {
                if (typeDef instanceof ComponentType) {
                    Map<String, NamedElement> myMap = new HashMap<String, NamedElement>();
                    myMap.put(this.typeDefinition, (NamedElement) typeDef);
                    myMap.put(this.nodeName, (NamedElement) containerNode);
                    results.add(myMap);
                }
            }
        }
        return results;
    }

    public String getScript(Map<String, NamedElement> myMap) {
        String script = templateScript;
        for (String name : myMap.keySet()) {
            String replacedString = "${" + name + "}";
            script = script.replace(replacedString, myMap.get(name).getName());
        }
        // the next line should be removed if we enable script without specifying componentName
        script = script.replace("${" + this.componentName + "}", myMap.get(this.typeDefinition).getName()+increment++ );
        return script;
    }

    @Override
    public org.kevoree.tools.marShell.ast.Script getASTScript(Map<String, NamedElement> stringNamedElementMap) {
        return null;  //To change body of implemented methods use File | Settings | File Templates.
    }

}
